descriptions.go

```diff
@@ -126,11 +131,255 @@ type RequestorInfo struct {
+		LogoPath *string                   `json:"logoPath,omitempty"`
+		Issues   *CredentialTypeIdentifier `json:"issues,omitempty"`
+
+		Info *TranslatedString `json:"info,omitempty"`
+		FAQ  []IssueWizardQA   `json:"faq,omitempty"`
+
+		Intro              *TranslatedString   `json:"intro,omitempty"`
+		SuccessHeader      *TranslatedString   `json:"successHeader,omitempty"`
+		SuccessText        *TranslatedString   `json:"successText,omitempty"`
+		ExpandDependencies *bool               `json:"expandDependencies,omitempty"`
+		Contents           IssueWizardContents `json:"contents"`
+	}
+
+	IssueWizardQA struct {
+		Question TranslatedString `json:"question"`
+		Answer   TranslatedString `json:"answer"`
+	}
+
+	IssueWizardContents [][][]IssueWizardItem
+
+	IssueWizardItem struct {
+		Type       IssueWizardItemType       `json:"type"`
+		Credential *CredentialTypeIdentifier `json:"credential,omitempty"`
+		Header     *TranslatedString         `json:"header,omitempty"`
+		Text       *TranslatedString         `json:"text,omitempty"`
+		Label      *TranslatedString         `json:"label,omitempty"`
+		SessionURL *string                   `json:"sessionUrl,omitempty"`
+		URL        *TranslatedString         `json:"url,omitempty"`
+		InApp      *bool                     `json:"inapp,omitempty"`
+	}
+
+	IssueWizardItemType string
+)
+
+const (
+	IssueWizardItemTypeCredential IssueWizardItemType = "credential"
+	IssueWizardItemTypeSession    IssueWizardItemType = "session"
+	IssueWizardItemTypeWebsite    IssueWizardItemType = "website"
+)
+
+// Choose from the wizard a list of items.
+//
+// If the ExpandDependencies boolean is set to false, the result of IssueWizardContents.Choose
+// is returned. If not set or set to true, this is augmented with all dependencies of all items
+// in an executable order.
+func (wizard IssueWizard) Choose(conf *Configuration, creds CredentialInfoList) ([]IssueWizardItem, error) {
+	// convert creds slice to map for easy lookup
+	credsmap := map[CredentialTypeIdentifier]struct{}{}
+	for _, cred := range creds {
+		credsmap[cred.Identifier()] = struct{}{}
+	}
+
+	contents := wizard.Contents.Choose(conf, credsmap)
+	if wizard.ExpandDependencies != nil && !*wizard.ExpandDependencies {
+		return contents, nil
+	}
+
+	// Each item in contents refers to a credential type that has dependencies, which may themselves
+	// have dependencies. So each item has a tree of dependencies. We must return a list
+	// containing all dependencies of each item in an executable order, i.e. item n in the
+	// list depends only on items < n in the list. We do this as follows:
+	// - by considering element n in items to be dependendent on element n-1, we join all
+	//   dependency trees into one
+	// - of that tree, starting at the leaf nodes and iterating downwards toward the root,
+	//   we put all items in the result list.
+
+	// Fist reverse the array, as we must start iterating at the root (having no dependants)
```

Comment:
I don't understand why reversing the array would not be necessary if all credential dependencies of each credential can be found in the scheme. The reason is that when iterating over the levels of the tree using`wizardItemVisit`, the iteration must start either at level 0 of the tree (i.e. at the root), or at the highest level of the tree that still contains items, but before the iteration itself that level is not yet known. So the only sensible starting point remaining is the root of the tree, so we reverse the wizard contents to give it index 0 instead of index `len(contents)-1`. This would be so even if all credential dependencies can be found in the scheme. (Or did I misunderstand your comment?)

Reply:
I was taking about the `reversed` variable. This is simply a reversed representation of the wizard items if I'm right. If all dependencies are known, the order of wizard items should not matter. Then the algorithm should always be able to find an order that works (unless there is a dependency cycle of course). So then it should not matter when it is the normal order or the reversed order.

The only reason why it matters is that when not all dependencies are in the scheme, you should be able to enforce an order by putting the wizard items in the right order.

Reply:
I have added a comment that should make it more clear.